### PR TITLE
Audio atom - add milestone events, time seeking and ad-free compatibility

### PR DESCRIPTION
**Pull request by JustinPinner**
Summary of changes for the audio atom;

- Now emits the milestone events for 25, 50 and 75%, and THE_END during playback
- Can be instructed to provide ad-free or Acast-wrapped audio source URLs
- Responds to user interaction with the progress bar, playing from the time selected (see graphic)

**Note: this version still carries the DEV version number, so don't merge until that's updated with the latest released version!**

Screenshots;
(captured locally) - there's a long-ish delay at 00:07:07 when the audio is first played - this is currently a known issue with audio playback

audio-atom-with-extra-scrubbability

### CODE REVIEW

File: core/src/main/resources/__flow__/types/ophan.fjs [hunk 1]
```diff
@@ -50,7 +50,12 @@ declare type Action = 'INSERT'
                     | 'VOTE'
                     | 'CLICK'
                     | 'PLAY'
-                    | 'READY';
+                    | 'READY'
+                    | 'PAUSE'
+                    | 'PERCENT25'
+                    | 'PERCENT50'
+                    | 'PERCENT75'
+                    | 'THE_END'
```
[the remark this answers is not in the document]
Comment (JustinPinner) [reply core/src/main/resources/__flow__/types/ophan.fjs]: 🤦‍♂️

File: core/src/main/resources/lib/audio.fjs [hunk 2]
```diff
@@ -92,48 +52,120 @@ export default (componentType: ComponentType) => ({ ophan, dom, viewport }: Serv
...
+    const now = Date.now();
+    const last = p.player.getAttribute('last-tick');
+
+    // throttle to 1 update per second
+    if (now - (Number(last) || 0) < 1000) return;
+
+    const played = Math.floor(p.player.currentTime);
+    const duration = Number(p.player.getAttribute('data-duration')) || 1; // avoid divide by zero if not present
+    const percentPlayed = Math.floor(played / duration * 100);
+    const milestones = p.player.getAttribute('milestones');
```
Comment (regiskuckaertz) [inline core/src/main/resources/lib/audio.fjs]: Just to be on the safe-side, can you please use `data-milestones` and `data-last-tick` instead?

File: core/src/main/resources/lib/audio.fjs [hunk 3]
```diff
@@ -97,43 +61,125 @@ export default (componentType: ComponentType) => ({ ophan, dom, viewport }: Serv
...
+
+    // throttle to 1 update per second
+    if (now - (Number(last) || 0) < 1000) return;
+
+    const played = Math.floor(audio.player.currentTime);
+    const duration = Number(audio.player.getAttribute('data-duration')) || 1; // avoid divide by zero if not present
+    const percentPlayed = Math.floor(played / duration * 100);
+    const milestones = audio.player.getAttribute('milestones');
+
+    switch (percentPlayed){
```
[the remark this answers is not in the document]
Comment (JustinPinner) [reply core/src/main/resources/lib/audio.fjs]: Thanks for the example. I've decided to leave this running with the `if` statements for the time being, as I'm not yet convinced that `fromEvent` improves legibility / clarity.

File: core/src/main/resources/lib/audio.fjs (hunk 2 above)
Comment (JustinPinner) [reply core/src/main/resources/lib/audio.fjs]: That should be done now. I did miss one though, and just committed it on its own.

File: core/src/main/resources/lib/audio.fjs [hunk 4]
```diff
@@ -97,43 +61,125 @@ export default (componentType: ComponentType) => ({ ophan, dom, viewport }: Serv
...
-      setPausedState(dom, p.playPauseButton);
+      setPausedState(p.playPauseButton);
       p.player.pause();
     }
   };
 
   const onTick = (p: Audio) => (a: Action): void => {
-    //TODO: send progress events to Ophan - needs thought re deduplication because it's based on percentage played
-    updateTime(dom, p);
+    updateTime(p);
```
[the remark this answers is not in the document]
Comment (JustinPinner) [reply core/src/main/resources/lib/audio.fjs]: Gone!

File: core/src/main/scala/Configuration.scala [hunk 5]
```diff
@@ -9,6 +9,8 @@ case object NilConfiguration extends NilConfiguration
 
 final case class ArticleConfiguration(
   ajaxUrl: String,
+  isAdFree: Boolean = false,
+  useAcast: Boolean = false,
```
[the remark this answers is not in the document]
Comment (JustinPinner) [reply core/src/main/scala/Configuration.scala]: I've simplified this a bit, so hopefully it's good enough now. Essentially the clients are responsible for applying the appropriate setting.

**Pull request merged**